engine/access/rest/common_query_params.go

```diff
@@ -0,0 +1,57 @@
+package rest
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+)
+
+type ctxKeyType string
+
+const expandQueryParam = "expand"
+const selectQueryParam = "select"
+
+// commonQueryParamMiddleware generates a Middleware function that extracts the given query parameter from the request
+// and adds it to the request context as a key value pair with the key as the query param name.
+// e.g. for queryParamName "fields", if the request url contains <some url>?fields=field1,field2,..fieldN,
```

Comment:
Are you sure you want to implement field and expand here ?

I think those should be another layer (some another real middleware), which can act as a proxy to this service. 

Most of the data you have here can be cached ( in redis for example ) on middleware for very long time. ( even the cache can be invalidated on demand from the engine, though it is a bit more advanced )

But in the end, selecting and expanding from cached result ( as a lot of the data is final ) would be much more performant

Reply:
Also for example if someone gets `getBlock` response for Block X , another one requests `getBlockHeader` for Block X, you will already have the required information from cache.

Reply:
so I would like the base implementation (this) to be able to serve the full api without the need of an external middleware. What you suggested could still be achieve with the a middleware in place, couldn't it?

Reply:
I think then it would be better to make select and expand full middleware, so they would manipulate data afterwards. 

IMO middleware should be pluggable ( and independent) in general. Maybe something like below: 

- Middleware: Cache lookup ( execute if not cached )  
- Middleware: Manipulate Response for selecting fields 
- Middleware: Manipulate Response for expand / collapse fields 
- Middleware: Output json from result 
etc etc 

So later this will allow to move some parts outside if necessary as parts will be loosely couples. 

PS: This is not too critical actually as soon AN will have full state.

Reply:
> I think then it would be better to make select and expand full middleware, so they would manipulate data afterwards.
> 
> IMO middleware should be pluggable ( and independent) in general. Maybe something like below:
> 
> * Middleware: Cache lookup ( execute if not cached )
> * Middleware: Manipulate Response for selecting fields
> * Middleware: Manipulate Response for expand / collapse fields
> * Middleware: Output json from result
>   etc etc
> 
> So later this will allow to move some parts outside if necessary as parts will be loosely couples.
> 
> PS: This is not too critical actually as soon AN will have full state.

Since the spec defines `select` and `expand` and since there is no middleware currently in place, I wanted to make sure that the implementation is fully compliant with the spec. When a middleware is in place, it can choose to not send the `select` and `expand` fields to this implementation and instead handle the expand and select logic by itself as you described. However, shouldn't the out-of-the-box implementation of an AN (this one) still implement select and expand?

Reply:
Yeah if you want full out-of-the-box spec compatibility, it is better to separate them in middlewares. 

What I meant here in this context was, you don't need middleware to extract query parameters. Instead you can put select and expand middlewares last. 

"Select Middleware" will select fields from response and filter, before it runs, you can cache results without filtering etc.

So if you remove "Select Middleware" everything works (except select function ofc), so each middleware Is making one job, instead of passing variables over context.

Reply:
Yeah actually that why  it was my first thought to scale this outside of AN. Currently we have with grpc everything is expanded right?

I think if there will be some good caching (last few blocks etc could be cached fully) it can decrease load a lot. 

Even I think as script execution will be on AN probably ( as state is going to sync with AN ), API should be extended with storage related functions too.  ( Considering AuthAccount for scripts on cadence just landed ) 

I foresee script execution will have very big load after this. Instead of event fetching, querying storage will be the new black :)